File: Quick_Sort/Quick_Sort.cpp
#include <iostream>

using namespace std;

int Parition(int array[], int left, int right)
{
    int pivot = array[left];
    int index = right;
    int temp;

    for(int j = right; j > left; j--)
    {
        if(array[j] > pivot)
        {
            temp = array[j];
            array[j] = array[index];
            array[index] = temp;
            index--;
        }
    }

    array[left] = array[index];
    array[index] = pivot;
    return index;
}

void Quick(int array[], int left, int right)
{
    if(left < right)
    {
        int pivot = Parition(array, left, right);

        Quick(array, left, pivot - 1);
        Quick(array, pivot + 1, right);
    }
}

void Quick_Sort(int array[], int size)
{
    Quick(array, 0, size - 1);
}

// Function to print elements of array
void Print_Array(int array[], int size)
{
    for(int i = 0; i < size; i++)
        cout << array[i] << " ";

    cout << endl;
}

int main()
{
    int array[] = {2, 4, 3, 1, 6, 8, 4};
    Quick_Sort(array, 7);
    Print_Array(array, 7);
    return 0;
}

// Output
// 1 2 3 4 4 6 8

File: Quick_Sort/Quick_Sort.ts
// Description: Function to perform the Quick Sort algorithm on an array of numbers
// Expected Output: returns the sorted array

// Helper function to get the Pivoting Element Index
function partition(array: number[], left: number = 0, right: number = array.length - 1) 
{
    const pivot = array[Math.floor((right + left) / 2)];
    let i = left;
    let j = right;

    while (i <= j) 
    {
        while (array[i] < pivot) 
        {
            i++;
        }

        while (array[j] > pivot)
        {
            j--;
        }

        // Swap values based on indices
        if (i <= j) 
        {
            [array[i], array[j]] = [array[j], array[i]];
            i++;
            j--;
        }
    }

    return i;
}


// Function that uses recursive definition for quick sort implementation
function quickSort(array: number[], left: number = 0, right: number = array.length - 1) 
{
    let index;

    if (array.length > 1)
    {
        index = partition(array, left, right);

        if (left < index - 1) 
        {
            quickSort(array, left, index - 1);
        }

        if (index < right) 
        {
            quickSort(array, index, right);
        }
    }

    return array;
}

// I/P
var arr = [6, 5, 4, 3, 2, 1];

console.log("Before sorting ",arr)
var res = quickSort(arr)

console.log("After sorting ",res) 
// O/P
// 1,2,3,4,5,6
